pkg/telemetryquerylanguage/contexts/tqllogs/logs.go

```diff
@@ -27,21 +27,21 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/telemetryquerylanguage/tql"
 )
 
-type LogTransformContext struct {
-	Log                  plog.LogRecord
+type TransformContext struct {
+	LogRecord            plog.LogRecord
```

Comment:
Why public fields when you have getters? Same question everywhere.

Reply:
The transform context is what processors use to identify which piece of data is going to be worked on with the next call to the tql conditions/functions.  Each context's TransformContext is an implementation of the `tql.TransformContext` interface, which require the `GetItem`, `GetInstrumentationScope`, and `GetResource` functions.  These functions are used within each context's `ParsePath` function, which the TQL will use to map from a Path to the underlying telemetry. The `GetSetter` that the `ParsePath` must return uses the generic `tql.TransformContext` interface.  Hence the need for `GetItem`, `GetInstrumentationScope`, and `GetResource`.

But the processors need a way to setup the TransformContext struct for its context.  There aren't any "Setters", so the field are set directly, and updated as the data is processed.  [Here is what it looks like for logs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6b553b94fb4e08123f9cbba7680127fb25d423af/processor/transformprocessor/internal/logs/processor.go#L45):
```
func (p *Processor) ProcessLogs(_ context.Context, td plog.Logs) (plog.Logs, error) {
	ctx := tqllogs.LogTransformContext{}
	for i := 0; i < td.ResourceLogs().Len(); i++ {
		rlogs := td.ResourceLogs().At(i)
		ctx.Resource = rlogs.Resource()
		for j := 0; j < rlogs.ScopeLogs().Len(); j++ {
			slogs := rlogs.ScopeLogs().At(j)
			ctx.InstrumentationScope = slogs.Scope()
			logs := slogs.LogRecords()
			for k := 0; k < logs.Len(); k++ {
				log := logs.At(k)
				ctx.Log = log

				for _, statement := range p.queries {
					if statement.Condition(ctx) {
						statement.Function(ctx)
					}
				}
			}
		}
	}
	return td, nil
}
```

If the fields weren't exposed (or equivalent "Setters" are not provided, which they currently are not) then the processor would not be able to set the transform context that is about to be run through each of the functions.

Reply:
Let's add a "NewTransformContext" which takes all the arguments and returns a `TransformContext`. This way we have the `TransformContext` as immutable so we don't have to worry about concurrency.

Reply:
Added.